### PR TITLE
fixes #591 - Handle zip/tgz/tar.gz (compressed files) with !syntax to identify single files

### LINKED ISSUE #591: Handle zip/tgz/tar.gz (compressed files) with !syntax to identify single files
apoc.load.json("file://foo.zip!foo/bar.json")

### PR DESCRIPTION
**Pull request by AngeloBusato**

### CODE REVIEW

File: src/main/java/apoc/util/Util.java [hunk 1]
```diff
@@ -281,32 +282,68 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
...
+
         if ("gzip".equals(encoding) || urlAddress.endsWith(".gz")) {
-            stream = new GZIPInputStream(stream);
+             return new GZIPInputStream(stream);
         }
         if ("deflate".equals(encoding)) {
-            stream = new DeflaterInputStream(stream);
+            return new DeflaterInputStream(stream);
         }
-        return new CountingInputStream(stream, sc.getLength());
```
Comment (jexp) [inline src/main/java/apoc/util/Util.java]: as you remove CountingInputStream here, did you fix that in all places that call: `getInputStream` ?

File: src/main/java/apoc/util/Util.java [hunk 2]
```diff
@@ -281,32 +282,68 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
 
     public static CountingInputStream openInputStream(String urlAddress, Map<String, Object> headers, String payload) throws IOException {
         StreamConnection sc;
+        InputStream stream;
+        if (urlAddress.contains(".zip") || urlAddress.contains(".tar") || urlAddress.contains(".tgz")) {
```
Comment (jexp) [inline src/main/java/apoc/util/Util.java]: shouldn't we do a `contains("!")` check first here
or at least check `zipFileName` for null before passing it into `getFileStreamIntoCompressedFile`

File: src/main/java/apoc/util/Util.java [hunk 3]
```diff
@@ -281,32 +282,68 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
...
-            stream = new DeflaterInputStream(stream);
+            return new DeflaterInputStream(stream);
         }
-        return new CountingInputStream(stream, sc.getLength());
+
+        return stream;
+    }
+
+    private static InputStream getFileStreamIntoCompressedFile(InputStream is, String fileName) throws IOException {
+        try (ZipInputStream zip = new ZipInputStream(is)) {
```
Comment (jexp) [inline src/main/java/apoc/util/Util.java]: there is no null check on `fileName` !!!

**Comment by jexp** [comment]
Thanks for adding all the tests.

File: src/main/java/apoc/util/Util.java [hunk 4]
```diff
@@ -281,32 +282,68 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
...
+            sc = getStreamConnection(urlAddress,headers,payload);
+            stream = getFileStreamIntoCompressedFile(sc.getInputStream(), zipFileName);
+
+            return new CountingInputStream(stream, sc.getLength());
+        }
+
+        sc = getStreamConnection(urlAddress,headers,payload);
+        stream = getInputStream(sc,urlAddress);
+
+        return new CountingInputStream(stream, sc.getLength());
```
Comment (AngeloBusato) [inline src/main/java/apoc/util/Util.java]: @jexp is here the CountingInputStream return.

File: src/main/java/apoc/util/Util.java (hunk 1 above)
Comment (AngeloBusato) [reply src/main/java/apoc/util/Util.java]: this is old, I done a small refactoring, i put I comment where is the new return countingInputStream, is inside the openInputStream method as before

**Comment by AngeloBusato** [comment]
@jexp PR updated

File: src/main/java/apoc/util/Util.java [hunk 5]
```diff
@@ -281,32 +282,72 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
 
     public static CountingInputStream openInputStream(String urlAddress, Map<String, Object> headers, String payload) throws IOException {
         StreamConnection sc;
+        InputStream stream;
+        if (urlAddress.contains("!")) {
+            String[] tokens = urlAddress.split("!");
+            urlAddress = tokens[0];
+            String zipFileName;
+            
+            if(tokens.length > 1) {
```
Comment (jexp) [inline src/main/java/apoc/util/Util.java]: ==2

File: src/main/java/apoc/util/Util.java [hunk 6]
```diff
@@ -281,32 +282,72 @@ private static String handleRedirect(URLConnection con, String url) throws IOExc
 
     public static CountingInputStream openInputStream(String urlAddress, Map<String, Object> headers, String payload) throws IOException {
         StreamConnection sc;
+        InputStream stream;
+        if (urlAddress.contains("!")) {
```
Comment (jexp) [inline src/main/java/apoc/util/Util.java]: this whole branch should go into a separate, properly named method

**Comment by AngeloBusato** [comment]
@jexp PR Updated

**Comment by jexp** [comment]
Thanks a lot.

**Pull request merged**